### PR TITLE
Performance improvements for the collision pipeline

### PR DESCRIPTION
**Pull request by nvtw**
## Description

Improves the performance of several collision detection kernels, mainly the sdf collisions. 
Introduces texture backed sparse block SDFs in order to work towards full CPU support of the collision pipeline (first step). They also bring a nice speedup on the GPU.
Most of the new code making the MR quite big is the new block sparse texture implementation and all the tests that come with it. The NanoVDB volumes can still be constructed but are currently mainly used for unit testing (they serve as ground truth when validating the block sparse texture backed sdfs).

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (if user-facing change)

## Summary by CodeRabbit

* **New Features**
  * Texture-backed sparse SDF system with selectable quantization (32/16/8-bit) and GPU textures
  * Models now expose texture-SDF resources and texture-based isomesh extraction

* **Improvements**
  * Faster, CPU/GPU-compatible mesh-mesh collisions using texture SDF sampling
  * Dynamic GPU load balancing and enhanced contact reduction for large collisions
  * Hydroelastic/contact generation and viewer isomesh caching updated for texture SDFs

* **Tests**
  * New comprehensive CUDA-focused tests for texture SDF functionality

* **Other**
  * Example: brick friction increased (0.5 → 0.7)

### CODE REVIEW

File: newton/_src/sim/model.py [hunk 1]
```diff
@@ -278,19 +278,23 @@ def __init__(self, device: Devicelike | None = None):
...
 
+        # Texture SDF storage
+        self.texture_sdf_data = None
+        """Compact array of TextureSDFData structs, shape [num_sdfs]."""
+        self.texture_sdf_coarse_textures = []
+        """Coarse 3D textures matching texture_sdf_data by index. Kept for reference counting."""
+        self.texture_sdf_subgrid_textures = []
+        """Subgrid 3D textures matching texture_sdf_data by index. Kept for reference counting."""
+        self.texture_sdf_subgrid_start_slots = []
+        """Subgrid start slot arrays matching texture_sdf_data by index. Kept for reference counting."""
```
[the remark this answers is not in the document]
Comment (nvtw) [reply newton/_src/sim/model.py:314]: @adenzler-nvidia @eric-heiden How do we deal with backwards compatibility/deprecation here? It's too expensive to keep the textures as nanoVDB and sparse Texture3D in memory

File: newton/_src/geometry/sdf_texture.py [hunk 2]
```diff
@@ -0,0 +1,1860 @@
...
+    points_np = mesh.points.numpy()
+    mesh_min = np.min(points_np, axis=0)
+    mesh_max = np.max(points_np, axis=0)
+
+    min_ext = mesh_min - margin
+    max_ext = mesh_max + margin
+
+    # Compute grid dimensions (same math as the former build_dense_sdf)
+    ext = max_ext - min_ext
+    max_ext_scalar = np.max(ext)
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/sdf_texture.py:1194]: **Division by zero for degenerate meshes**: If `max_ext_scalar` is zero (e.g., flat planar mesh), `cell_size_scalar` is zero, and `ext / cell_size_scalar` produces NaN. `np.ceil(nan).astype(int)` is implementation-defined.

Similarly at line 1127, `1.0 / float(cell_size[i])` produces infinity when any extent is zero.

Suggestion: add early validation like `if max_ext_scalar < 1e-10: return create_empty_texture_sdf_data(device)` or raise a `ValueError`.

File: newton/_src/geometry/sdf_texture.py [hunk 3]
```diff
@@ -0,0 +1,1860 @@
...
+    """Sparse SDF stored in 3D CUDA textures with indirection array.
+
+    Uses a two-level structure:
+    - A coarse 3D texture for background/far-field sampling
+    - A packed subgrid 3D texture for narrow-band high-resolution sampling
+    - An indirection array mapping coarse cells to subgrid texture blocks
+    """
+
+    # Textures and indirection
+    coarse_texture: wp.Texture3D
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/sdf_texture.py:70]: Nit: The sentinel values `0xFFFFFFFF` (empty subgrid) and `0xFFFFFFFE` (linearized subgrid) appear as raw literals in 10+ locations across construction and sampling code. A typo would silently corrupt sampling.

Suggestion: define module-level constants:
```python
SLOT_EMPTY = np.uint32(0xFFFFFFFF)
SLOT_LINEAR = np.uint32(0xFFFFFFFE)
```

File: newton/_src/geometry/sdf_contact.py [hunk 4]
```diff
@@ -705,106 +441,356 @@ def get_triangle_count(shape_type: int, mesh_id: wp.uint64, hfd: HeightfieldData
...
 
-    # Precompute SDF AABB in unscaled local space for the volume path.
-    # For BVH and heightfield paths, the per-sample functions already have
-    # their own distance culling so we skip the AABB pre-filter.
-    sdf_aabb_lower = sdf_data.center - sdf_data.half_extents
-    sdf_aabb_upper = sdf_data.center + sdf_data.half_extents
+        difference = wp.max(difference, 1e-8)
 
-    synchronize()  # Ensure buffer state is consistent before starting
+        tolerance_sq = 3e-3 * 3e-3
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/sdf_contact.py:538]: `tolerance_sq` changed from `1e-3 * 1e-3` (1e-6) to `3e-3 * 3e-3` (9x looser). If this is intentional to compensate for the texture SDF's different accuracy characteristics vs NanoVDB, a brief comment explaining why would help future readers.

File: newton/_src/geometry/sdf_contact.py [hunk 5]
```diff
@@ -705,106 +441,356 @@ def get_triangle_count(shape_type: int, mesh_id: wp.uint64, hfd: HeightfieldData
...
-            v2 = wp.cw_mul(v2_scaled, inv_sdf_scale)
-            bounding_sphere_center, bounding_sphere_radius = get_bounding_sphere(v0, v1, v2)
+                if use_bvh_for_sdf:
+                    _, sdf_gradient = sample_sdf_grad_using_mesh(sdf_mesh_id, p)
+                else:
+                    _, sdf_gradient = texture_sample_sdf_grad(texture_sdf, p)
+
+            grad_len = wp.length(sdf_gradient)
+            if grad_len == 0.0:
+                sdf_gradient = wp.vec3(0.571846586, 0.705545099, 0.418566116)
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/sdf_contact.py:560]: Nit: this magic fallback gradient appears to be an arbitrary non-axis-aligned unit vector for when the SDF gradient is zero. A brief comment explaining the choice (e.g., "arbitrary unit vector to break symmetry") would help.

Also: further down at line ~1198, when `direction_len == 0.0` the entire contact is silently dropped even though the SDF detected a collision. Consider falling back to center-to-contact vector as the normal instead of discarding the contact.

File: newton/_src/geometry/multicontact.py [hunk 6]
```diff
@@ -930,74 +831,92 @@ def build_manifold(
...
 
-        # Compute center for 2D projection
         center = 0.5 * (p_a + p_b)
 
-        # Allocate buffers for contact polygons (2D projected)
-        b_buffer = wp.zeros(shape=(12,), dtype=wp.vec2f)
-        # a_buffer = wp.array(ptr=b_buffer.ptr + wp.uint64(6 * 8), shape=(6,), dtype=wp.vec2f)
-        a_buffer = wp.array(ptr=get_ptr(b_buffer) + wp.uint64(6 * 8), shape=(6,), dtype=wp.vec2f)
+        # Allocate buffers: 5 for A, up to 10 for B (5 + clipping headroom)
+        b_buffer = wp.zeros(shape=(10,), dtype=wp.vec2f)
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/multicontact.py:858]: **Potential buffer aliasing issue**: `a_buffer` starts at `b_buffer[5]` (aliased memory). During Sutherland-Hodgman clipping, `trim_all_in_place` writes to `b_buffer` (m_b), which can grow up to 10 elements. Writes to `b_buffer[5..9]` would overwrite `a_buffer` entries while they may still be needed as the read-only clipping polygon.

The previous code used 12 for b_buffer and 6 for a_buffer. With 5 polygon points from A and 5 from B, worst case is exactly 10 — which is tight. Could you verify that `a_buffer` is fully consumed before `b_buffer` overwrites those slots, or add padding?

File: newton/_src/geometry/contact_reduction_global.py [hunk 7]
```diff
@@ -802,6 +805,124 @@ def export_and_reduce_contact(
...
+    """
+    ht_capacity = reducer_data.ht_capacity
+    use_beta = depth < wp.static(BETA_THRESHOLD)
+
+    # === Normal bin: find/create hashtable entry ===
+    bin_id = get_slot(normal)
+    pos_2d = project_point_to_plane(bin_id, centered_position)
+    key = make_contact_key(shape_a, shape_b, bin_id)
+
+    entry_idx = hashtable_find_or_insert(key, reducer_data.ht_keys, reducer_data.ht_active_slots)
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/contact_reduction_global.py:861]: **`ht_insert_failures` counter never incremented**: When `entry_idx < 0` (hashtable full), the contact skips reduction silently. The `ht_insert_failures` field exists on the struct and is zeroed, but never incremented here or at the other `hashtable_find_or_insert` call sites (~line 890, 918). This makes the only diagnostic for hashtable overflow non-functional.

Suggestion: add `wp.atomic_add(reducer_data.ht_insert_failures, 0, 1)` in the `entry_idx < 0` branches.

File: newton/_src/sim/builder.py [hunk 8]
```diff
@@ -9570,85 +9598,109 @@ def compute_voxel_resolution_from_aabb(aabb_lower, aabb_upper, voxel_budget):
...
+                                compact_texture_sdf_subgrid_start_slots.append(None)
+                        else:
+                            prim_mesh = _create_primitive_mesh(shape_type, shape_scale)
+                            if prim_mesh is not None:
+                                prim_wp_mesh = wp.Mesh(
+                                    points=wp.array(prim_mesh.vertices, dtype=wp.vec3, device=device),
+                                    indices=wp.array(prim_mesh.indices.flatten(), dtype=wp.int32, device=device),
+                                    support_winding_number=True,
+                                )
+                                tex_data, c_tex, s_tex, tex_bc = create_texture_sdf_from_mesh(
```
Comment (adenzler-nvidia) [inline newton/_src/sim/builder.py]: **No error handling for texture SDF construction**: If `create_texture_sdf_from_mesh` fails (GPU OOM, CUDA texture dimension limits, degenerate mesh), the exception crashes the entire model build with no context about which shape failed.

Suggestion: wrap in try-except, log a warning identifying the shape, and fall back to `create_empty_texture_sdf_data()` so the model still builds (with BVH fallback for that shape).

**Review by adenzler-nvidia — COMMENTED** [review]
## PR Review Summary

This is a well-designed PR with a solid texture SDF architecture, clean fused MPR+GJK approach, and good test coverage for the new subsystem. The main concerns are around edge-case robustness (degenerate meshes, buffer overflows) and a few test coverage gaps.

**Warp version**: All texture APIs used (`wp.Texture3D`, `wp.texture_sample` with `dtype=float`, `wp.TextureFilterMode.LINEAR`) are available in Warp 1.12.0. No version bump needed.

### Critical Issues (3)
1. **Division by zero** with degenerate mesh scales/extents — `min_sdf_scale` in `sdf_contact.py:892,1124` used as divisor without guard; `cell_size` in `sdf_texture.py:1127` and `max_ext_scalar` in `sdf_texture.py:1189` (see inline comments)
2. **Buffer aliasing** in `multicontact.py` — `a_buffer` overlaps `b_buffer[5:]`, clipping may corrupt read-only polygon (see inline)
3. **`ht_insert_failures` never incremented** in `contact_reduction_global.py` — diagnostic counter is allocated but useless (see inline)

### Important Issues (7)
4. `tolerance_sq` tripled (1e-6 → 9e-6) without explanation in `sdf_contact.py` (see inline)
5. Single-threaded prefix sum kernels (`dim=1`) in `sdf_contact.py:compute_mesh_mesh_block_offsets` and `narrow_phase.py:compute_mesh_plane_block_offsets` may bottleneck with many mesh pairs
6. Silent BVH fallback (`sdf_contact.py:854`) with no diagnostics when texture SDF is missing/invalid
7. No validation of `create_texture_sdf_from_mesh` failure in builder (see inline)
8. Silent contact buffer overflow (`contact_reduction_global.py:656`) drops contacts with no counter
9. Dual emptiness convention — `subgrid_size == 0` vs `coarse_texture.width == 0` checked in different places
10. Magic sentinel values `0xFFFFFFFF`/`0xFFFFFFFE` as raw literals in 10+ locations (see inline)

### Test Coverage Gaps (4)
11. No test for `export_and_reduce_contact_centered` (new pre-pruning logic in `contact_reduction_global.py`)
12. No test for `support_map_lean` / `use_lean_gjk_mpr` code path
13. Removed `ContactReductionFunctions` tests (189 lines) not replaced for mesh-plane reduction
14. `test_export_reduced_contacts_kernel` may have stale signature after `exported_flags` was added

### Additional Notes
- `sdf_contact.py:1198-1200`: Contact silently dropped when `direction_len == 0.0` — consider falling back to center-to-contact vector instead
- `sdf_contact.py:892,1124`: `min_sdf_scale` from user shape scale used as divisor — zero scale produces `inf` on GPU. Guard with `wp.max(min_sdf_scale, 1e-10)` or validate in builder
- `sdf_utils.py:367`: `tex_block_coords` may be unbound on CPU-only systems (`wp.is_cuda_available() == False`). Initialize to `None` before the branch

### Strengths
- Two-level coarse/subgrid texture SDF architecture well-suited for GPU hardware
- Analytical trilinear gradient from 8 texel reads (vs 56 for finite differences)
- Fused MPR+GJK sharing support functions and frame transforms reduces code generation
- Linearization optimization demoting near-linear subgrids is a smart memory tradeoff
- New `test_sdf_texture.py` (774 lines, 14 tests) is well-structured with statistical validation
- Dynamic load balancing via prefix-sum block allocation is correctly implemented

File: newton/_src/geometry/sdf_contact.py (hunk 4 above)
Comment (nvtw) [reply newton/_src/geometry/sdf_contact.py:538]: This leads to a quite significant per win and the I could not observe worse contact quality. I assume that to converge to the smaller tolerance, the float32 precision might have started to matter which can be harmful for convergence.

File: newton/_src/geometry/multicontact.py (hunk 6 above)
Comment (nvtw) [reply newton/_src/geometry/multicontact.py:858]: This is not an issue. It's by design. All algorithms work with the aliased buffers. It allows to save valuable gpu registers at no cost.

File: newton/_src/sim/builder.py (hunk 8 above)
Comment (nvtw) [reply newton/_src/sim/builder.py]: Should be improved now

File: newton/_src/geometry/contact_reduction_global.py (hunk 7 above)
Comment (nvtw) [reply newton/_src/geometry/contact_reduction_global.py:861]: Should be improved now

File: newton/_src/geometry/sdf_contact.py (hunk 5 above)
Comment (nvtw) [reply newton/_src/geometry/sdf_contact.py:560]: Should be improved now

File: newton/_src/geometry/sdf_texture.py (hunk 3 above)
Comment (nvtw) [reply newton/_src/geometry/sdf_texture.py:70]: Should be improved now

File: newton/_src/geometry/sdf_texture.py (hunk 2 above)
Comment (nvtw) [reply newton/_src/geometry/sdf_texture.py:1194]: Should be improved now

**Comment by nvtw** [comment]
> ## PR Review Summary
> This is a well-designed PR with a solid texture SDF architecture, clean fused MPR+GJK approach, and good test coverage for the new subsystem. The main concerns are around edge-case robustness (degenerate meshes, buffer overflows) and a few test coverage gaps.
> 
> **Warp version**: All texture APIs used (`wp.Texture3D`, `wp.texture_sample` with `dtype=float`, `wp.TextureFilterMode.LINEAR`) are available in Warp 1.12.0. No version bump needed.
> 
> ### Critical Issues (3)
> 1. **Division by zero** with degenerate mesh scales/extents — `min_sdf_scale` in `sdf_contact.py:892,1124` used as divisor without guard; `cell_size` in `sdf_texture.py:1127` and `max_ext_scalar` in `sdf_texture.py:1189` (see inline comments)
> 2. **Buffer aliasing** in `multicontact.py` — `a_buffer` overlaps `b_buffer[5:]`, clipping may corrupt read-only polygon (see inline)
> 3. **`ht_insert_failures` never incremented** in `contact_reduction_global.py` — diagnostic counter is allocated but useless (see inline)
> 
> ### Important Issues (7)
> 4. `tolerance_sq` tripled (1e-6 → 9e-6) without explanation in `sdf_contact.py` (see inline)
> 5. Single-threaded prefix sum kernels (`dim=1`) in `sdf_contact.py:compute_mesh_mesh_block_offsets` and `narrow_phase.py:compute_mesh_plane_block_offsets` may bottleneck with many mesh pairs
> 6. Silent BVH fallback (`sdf_contact.py:854`) with no diagnostics when texture SDF is missing/invalid
> ...
> * Two-level coarse/subgrid texture SDF architecture well-suited for GPU hardware
> * Analytical trilinear gradient from 8 texel reads (vs 56 for finite differences)
> * Fused MPR+GJK sharing support functions and frame transforms reduces code generation
> * Linearization optimization demoting near-linear subgrids is a smart memory tradeoff
> * New `test_sdf_texture.py` (774 lines, 14 tests) is well-structured with statistical validation
> * Dynamic load balancing via prefix-sum block allocation is correctly implemented

Most comments are now addressed. Not all of them are really required, so I was a bit selective.

**Comment by adenzler-nvidia** [comment]
### Code review

Found 2 issues:

1. `b_buffer` / `a_buffer` aliasing in `build_manifold` will corrupt the clipping polygon during Sutherland-Hodgman. `b_buffer` has 10 slots with `a_buffer` aliased at index 5 (byte offset `5 * 8`). Clipping a 5-vertex polygon B against 5 edges of A can produce up to 10 output vertices, meaning indices 5--9 overwrite `a_buffer[0..4]` mid-algorithm. The old 12-slot / offset-6 layout had headroom for the 6-vertex case; the new 10-slot / offset-5 layout has zero headroom for the 5-vertex case.

https://github.com/newton-physics/newton/blob/ced589dda0db5f39160a00982a7b7bdb8167b0ee/newton/_src/geometry/multicontact.py#L857-L860

2. `GlobalContactReducer` is still sized to `max_triangle_pairs` only, but now both mesh-mesh **and** mesh-plane contacts are routed through it. For scenes with dense meshes resting on planes, mesh-plane contacts alone can saturate the capacity, silently dropping contacts with no warning surfaced to the caller.

https://github.com/newton-physics/newton/blob/ced589dda0db5f39160a00982a7b7bdb8167b0ee/newton/_src/geometry/narrow_phase.py#L1654-L1656

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by adenzler-nvidia** [comment]
Two items from the earlier review don't appear to be addressed in the latest commits -- wanted to check whether these are non-issues or still on the list:

1. **`b_buffer` / `a_buffer` aliasing in `multicontact.py:858`**: `b_buffer` is 10 slots with `a_buffer` aliased at index 5. Sutherland-Hodgman clipping can grow `b_buffer` up to 10 entries, overwriting `a_buffer` mid-clip. Is the 5-vertex polygon guaranteed to never expand past 5 during clipping, or does this still need a fix?

2. **`GlobalContactReducer` capacity in `narrow_phase.py:1655`**: The reducer is sized to `max_triangle_pairs`, but now handles both mesh-mesh and mesh-plane contacts. Is mesh-plane volume small enough in practice that the existing capacity is sufficient, or should it be increased?

**Comment by adenzler-nvidia** [comment]
Minor: `Model.sdf_data`, `Model.sdf_volume`, and `Model.sdf_coarse_volume` are removed and replaced by `Model.texture_sdf_data`, `Model.texture_sdf_coarse_textures`, etc. All usage is internal (`_src/` and tests), so deprecation shims aren't necessary, but could you add a `Changed` entry to the CHANGELOG noting the replacement? Something like:

```
### Changed

- Replace `Model.sdf_data` / `sdf_volume` / `sdf_coarse_volume` with texture-based equivalents (`texture_sdf_data`, `texture_sdf_coarse_textures`, `texture_sdf_subgrid_textures`).
```

**Comment by nvtw** [comment]
> Two items from the earlier review don't appear to be addressed in the latest commits -- wanted to check whether these are non-issues or still on the list:
> 
> 1. **`b_buffer` / `a_buffer` aliasing in `multicontact.py:858`**: `b_buffer` is 10 slots with `a_buffer` aliased at index 5. Sutherland-Hodgman clipping can grow `b_buffer` up to 10 entries, overwriting `a_buffer` mid-clip. Is the 5-vertex polygon guaranteed to never expand past 5 during clipping, or does this still need a fix?
> 2. **`GlobalContactReducer` capacity in `narrow_phase.py:1655`**: The reducer is sized to `max_triangle_pairs`, but now handles both mesh-mesh and mesh-plane contacts. Is mesh-plane volume small enough in practice that the existing capacity is sufficient, or should it be increased?

1. The algorithm is designed that if the clipping appends elements, then then at least the same amount of elements has already been consumed from the second (aliased) half and they won't be read anymore. The algorithm was exactly designed for this - maximally efficient register use.
2. max_triangle_pairs has a default value of 1mio and can be user set if it is too small. So this should not be an issue as well.

Also this clipping algorithm (that uses the aliased buffers) was there before, it's not new. This MR just touches the surroundings of this algorithm a bit and reduces from 6 to 5 scan directions which should be sufficient (confirmed it by testing my sample scenes). It is worth trying it. Worst case is that we just revert back to 6 scan directions if users complain and that's not a bad worst case.

**Comment by nvtw** [comment]
> Minor: `Model.sdf_data`, `Model.sdf_volume`, and `Model.sdf_coarse_volume` are removed and replaced by `Model.texture_sdf_data`, `Model.texture_sdf_coarse_textures`, etc. All usage is internal (`_src/` and tests), so deprecation shims aren't necessary, but could you add a `Changed` entry to the CHANGELOG noting the replacement? Something like:
> 
> ```
> ### Changed
> 
> - Replace `Model.sdf_data` / `sdf_volume` / `sdf_coarse_volume` with texture-based equivalents (`texture_sdf_data`, `texture_sdf_coarse_textures`, `texture_sdf_subgrid_textures`).
> ```

Done

**Comment by adenzler-nvidia** [comment]
### Code review

Found 1 issue:

1. Missing `device` scoping for texture SDF construction in `create_from_mesh()`. The NanoVDB path passes `device=device`, but the new texture path allocates `wp.array`, `wp.Mesh`, and calls `create_texture_sdf_from_mesh()` without specifying a device or wrapping in `wp.ScopedDevice(device)`. In multi-GPU setups, these allocations will default to the current device rather than the intended one. This same class of bug was previously fixed in PR #1972.

https://github.com/newton-physics/newton/blob/76bcad22c6224a725bfe608f68d654742b0afd15/newton/_src/geometry/sdf_utils.py#L248-L270

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by nvtw** [comment]
> ### Code review
> Found 1 issue:
> 
> 1. Missing `device` scoping for texture SDF construction in `create_from_mesh()`. The NanoVDB path passes `device=device`, but the new texture path allocates `wp.array`, `wp.Mesh`, and calls `create_texture_sdf_from_mesh()` without specifying a device or wrapping in `wp.ScopedDevice(device)`. In multi-GPU setups, these allocations will default to the current device rather than the intended one. This same class of bug was previously fixed in PR [Fix device scoping issues in multi-GPU environments #1972](https://github.com/newton-physics/newton/pull/1972).
> 
> https://github.com/newton-physics/newton/blob/76bcad22c6224a725bfe608f68d654742b0afd15/newton/_src/geometry/sdf_utils.py#L248-L270
> 
> 🤖 Generated with [Claude Code](https://claude.ai/code)
> 
> - If this code review was useful, please react with 👍. Otherwise, react with 👎.

Done
